docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,381 @@
+    │   ├── defaults_test.go
+    │   ├── register.go
+    │   ├── types.generated.go
+    │   ├── types.go
+    │   └── types_swagger_doc_generated.go
+    └── validation
+        ├── validation.go
+        └── validation_test.go
+```
+
+## Desired directory structure
+
+We want to define a standard layout and structure for Kubernetes API types.
+Below is a first draft, with explanation afterwards. We may try to move the
+relevant parts of the legacy `pkg/api/v1` api into the new structure, although
+this api doesn't have a group.
+
+```
+apis
+└── group
+    ├── install
+    ├── types.go
+    ├── v1
+    │   ├── install
+    │   │   ├── conversion
+    │   │   │   ├── generated_conversion.go
+    │   │   │   └── manual_conversion_overrides.go
+    │   │   ├── defaulting
+    │   │   │   ├── generated_defaulters.go
+    │   │   │   └── generated_default_stubs.go
+    │   │   └── validation
+    │   │       ├── generated_validation.go
+    │   │       └── generated_validation_stubs.go
+    │   └── types.go
+    ├── v1beta1
+    │   ├── install
+    │   │   ├── conversion
+    │   │   │   ├── generated_conversion.go
+    │   │   │   └── manual_conversion_overrides.go
+    │   │   ├── defaulting
+    │   │   │   ├── generated_defaulters.go
+    │   │   │   └── generated_default_stubs.go
+    │   │   └── validation
+    │   │       ├── generated_validation.go
+    │   │       └── generated_validation_stubs.go
+    │   └── types.go
+    └── v2alpha1
+        ├── install
+        │   ├── conversion
+        │   │   ├── generated_conversion.go
+        │   │   └── manual_conversion_overrides.go
+        │   ├── defaulting
+        │   │   ├── generated_defaulters.go
+        │   │   └── generated_default_stubs.go
+        │   └── validation
+        │       ├── generated_validation.go
+        │       └── generated_validation_stubs.go
+        └── types.go
+```
+
+### Package: `apis/`
+
+Multiple groups may live in the same tree. Each group gets its own directory
+under `apis/`. We'll assume an `apis/` directory of this structure in all tools.
+
+TBD: Either test code will live here, or no code.
+
+### Package: `apis/group/`
+
+Multiple versions live in a group. Each version gets its own subdirectory.
+
+Additionally, a types.go file contains an "internal" version of the group's
+objects. This file is used as a destination/source for conversion functions.
+
+### Package: `apis/group/install`
+
+This package installs *all* of the versions of the group.
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+`types.go` contains the types for this version of the api group.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved to `pkg/apis/unversioned`) and
+little else. In particular, it *must not* import its parent `apis/group`.
+
+### Package: `apis/group/v1/install`
+
+This package, if imported, registers the v1 version of this group, its
```

Comment:
Is it a side-effect of import (`init()`) or an intentional call to a well-known function?  Maybe not a huge deal yet, but it would be nice it this only had intra-group-deps (i.e we it doesn't access global variables in the api pkg (Scheme, ahem)).